packages/@react-aria/utils/src/scrollIntoView.ts

```diff
@@ -63,3 +70,56 @@ function relativeOffset(ancestor: HTMLElement, child: HTMLElement, axis: 'left'|
   }
   return sum;
 }
+
+export function scrollIntoViewFully(target: Element, scrollOptions?: ScrollIntoViewOptions) {
+  if (scrollOptions == null) {
+    scrollOptions = {
+      block: 'nearest'
+    };
+  }
+
+  if (shouldScrollIntoView(target)) {
+    scrollIntoViewHelper(target, scrollOptions);
+  }
+}
+
+// Determines if a target element is fully out of view and thus should be scrolled into view if possible.
+// Recurses through every scrollable parent of the target and returns true the target is not visible in any of the scrollable parents.
+function shouldScrollIntoView(target: Element, originalTarget?: Element) {
+  let root = document.scrollingElement || document.documentElement;
+  if (target === root || target == null) {
+    return false;
+  }
+
+  // Note: there are cases where the target can be out of view but doesn't have a scrollable parent that would accurately determine
+  // if it is out of view (e.g. table resizer is absolutely positioned and can be moved out of view but the scrollable element that would bring it
+  // into view is the table's body rather than the column headers).
+  let scrollParent = getScrollParent(target);
+  let targetToCompare = originalTarget || target;
+  let {bottom: targetBottom, top: targetTop, left: targetLeft, right: targetRight} = targetToCompare.getBoundingClientRect();
+  let {bottom: scrollParentBottom, top: scrollParentTop, left: scrollParentLeft, right: scrollParentRight} = scrollParent.getBoundingClientRect();
+
+  if (scrollParent === root) {
+    let viewportHeight = window.visualViewport?.height ?? window.innerHeight;
+    let viewportWidth = window.visualViewport?.width ?? window.innerWidth;
+    return (targetBottom < 0 || targetTop > viewportHeight || targetLeft > viewportWidth || targetRight < 0);
+  }
+
+  return targetBottom <= scrollParentTop || targetTop >= scrollParentBottom || targetLeft >= scrollParentRight || targetRight <= scrollParentLeft || shouldScrollIntoView(scrollParent, targetToCompare);
+}
```

Comment:
I'm curious why this is preferable using `IntersectionObserver` as in https://github.com/adobe/react-spectrum/pull/2767/files#diff-2814ab64465ca7396b113c75472ed75d06046b918d1cc620e8225a7531c35c04R17-R47? In the ListBox example using `block: nearest` it seems like ListItems on the bottom of the viewport are not scrolling fully into view.

Reply:
There were concerns brought up about IntersectionObserver's performance impact vs just calculating if an element is in view. It felt like the IntersectionObserver was a bit overkill since it felt more geared towards things like infinite scrolling/observing the visibility of the element during its lifespan. I'll definitely do some testing for the performance between the two though.

As for the some ListBox items not scrolling into view, thanks for catching that, I'll take a look. It might be returning that those items are barely in view or something

Reply:
I also forgot to mention that IntersectionObserver will always return that a visually hidden element is not in view. This is expected but causes issues when we focus visually hidden elements like a slider's hidden input and perform keyboard operation on them. To a user, it seems as if the thumb of the slider is focused and thus is in view but to the scrollIntoView code it thinks we need to scroll the hidden input into view even though it is in the viewport. In the current PR this isn't an issue since I only scroll collection elements into view now but it would be an issue when we expand this scroll into view behavior to more components in followup work